tests/LanguageServerRename/Integration/Handler/RenameHandlerTest.php

```diff
@@ -0,0 +1,128 @@
+use Phpactor\TextDocument\TextEdit;
+use Phpactor\TextDocument\TextEdits;
+
+class RenameHandlerTest extends IntegrationTestCase
+{
+    const EXAMPLE_FILE = 'file:///Foobar.php';
+    const EXAMPLE_NEW_NAME = 'foobar';
+
+
+    /**
+     * @var LanguageServerTester
+     */
+    private $tester;
+
+    /**
+     * @var InMemoryRenamer
+     */
+    private $renamer;
+
+    public function testRegistersCapabilities(): void
+    {
+        $this->bootContainerWithRangeAndResults(null, []);
+        $result = $this->tester->initialize();
+        self::assertTrue($result->capabilities->renameProvider->prepareProvider);
+    }
+
+    public function testPrepareRenameReturnsNullIfItCouldNotPrepareAnything(): void
+    {
+        $this->bootContainerWithRangeAndResults(null, []);
+        $this->tester->textDocument()->open(self::EXAMPLE_FILE, '<?php');
+
+        $response = $this->tester->requestAndWait(
+            PrepareRenameRequest::METHOD,
+            new PrepareRenameParams(
+                ProtocolFactory::textDocumentIdentifier(self::EXAMPLE_FILE),
+                ProtocolFactory::position(0, 0),
+            )
+        );
+
+        $this->tester->assertSuccess($response);
+        self::assertNull($response->result);
+    }
+
+    public function testPrepareRename(): void
+    {
+        $expectedCharOffset = 3;
+
+        $this->bootContainerWithRangeAndResults(ByteOffsetRange::fromInts(0, $expectedCharOffset), [
+            new RenameResult(TextEdits::none(), TextDocumentUri::fromString('file:///foobar.php'))
+        ]);
+        $this->tester->textDocument()->open(self::EXAMPLE_FILE, '<?php');
+
+        $response = $this->tester->requestAndWait(
+            PrepareRenameRequest::METHOD,
+            new PrepareRenameParams(
+                ProtocolFactory::textDocumentIdentifier(self::EXAMPLE_FILE),
+                ProtocolFactory::position(0, 0),
+            )
+        );
+
+        $this->tester->assertSuccess($response);
+        self::assertEquals(ProtocolFactory::range(0, 0, 0, $expectedCharOffset), $response->result);
+    }
+
+    public function testRename(): void
+    {
+        $expectedUri = TextDocumentUri::fromString(self::EXAMPLE_FILE);
+        $this->bootContainerWithRangeAndResults(ByteOffsetRange::fromInts(0, 0), [
+            new RenameResult(
+                TextEdits::one(
+                    TextEdit::create(ByteOffset::fromInt(1), 0, self::EXAMPLE_NEW_NAME)
+                ),
+                $expectedUri,
+            )
+        ]);
+        $this->tester->textDocument()->open(self::EXAMPLE_FILE, '<?php');
+
+        $response = $this->tester->requestAndWait(
+            RenameRequest::METHOD,
+            new RenameParams(
+                ProtocolFactory::textDocumentIdentifier(self::EXAMPLE_FILE),
+                ProtocolFactory::position(0, 0),
+                self::EXAMPLE_NEW_NAME
+            )
+        );
+
+        $this->tester->assertSuccess($response);
```

Comment:
The tester has a utulity method to check for a successful result and throws any error it returns.